src/dynamic_diagnostic.rs

```diff
@@ -0,0 +1,72 @@
+use std::{
+    error::Error,
+    fmt::{Debug, Display},
+};
+
+use crate::Diagnostic;
+
+/// Diagnostic that can be created at runtime.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DynamicDiagnostic {
+    /// Displayed diagnostic description
+    pub description: String,
+    /// Unique diagnostic code to look up more information
+    /// about this Diagnostic. Ideally also globally unique, and documented
+    /// in the toplevel crate's documentation for easy searching.
+    /// Rust path format (`foo::bar::baz`) is recommended, but more classic
+    /// codes like `E0123` will work just fine.
+    pub code: Option<String>,
+}
+
+impl Display for DynamicDiagnostic {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", &self.description)
+    }
+}
+
+impl Error for DynamicDiagnostic {}
+
+impl Diagnostic for DynamicDiagnostic {
+    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        self.code
+            .as_ref()
+            .map(Box::new)
+            .map(|c| c as Box<dyn Display>)
+    }
+}
+
+impl DynamicDiagnostic {
+    /// Create a new dynamic diagnostic with the given description.
+    ///
+    /// # Examples
+    /// ```
+    /// use miette::{Diagnostic, DynamicDiagnostic};
+    ///
+    /// let diag = DynamicDiagnostic::new("Oops, something went wrong!");
+    /// assert_eq!(diag.to_string(), "Oops, something went wrong!");
+    /// assert_eq!(diag.description, "Oops, something went wrong!");
+    /// ```
+    pub fn new(description: impl Into<String>) -> Self {
```

Comment:
I think the usual pattern is AsRef<str>, no?

Reply:
You mean `AsRef<str>`? But this will be less efficient if argument is already an owned `String`